Add support for updating packages on Slurm nodes

priteau opened the pull request:
Slurm compute nodes are drained before applying updates. A reboot is triggered if a new kernel has been installed.

This commit also includes code to configure older OpenHPC repos, allowing to test an update of Slurm packages from 17.11.7 to 17.11.10 by passing configure_older_updates=true to the jasmin.yml playbook.

Note: the openhpc-repos.yml playbook is only for testing Slurm package updates and can eventually be removed.

priteau committed:
Add playbooks to drain and resume Slurm compute nodes

oneswig commented on drain.yml:
```diff
@@ -0,0 +1,20 @@
+---
+# Ansible playbook to drain Slurm compute nodes. Waits for compute nodes to be
```
Technically it's not a playbook because it doesn't select hosts or groups on which to apply the tasks.  But that's fine, just me being pedantic...

priteau replied:
You're right, I noticed it as I was writing the comment but only researched it later. In a role this would be a task file. I created a `tasks` directory for them.

oneswig commented on drain.yml:
```diff
@@ -0,0 +1,20 @@
+
+- name: Drain compute nodes
+  become: yes
+  shell: "scontrol update nodename={{ item }} state=DRAIN reason='maintenance'"
```
Consider using `command` in place of `shell` if only because the process group is slightly smaller.

priteau replied:
Thanks for the advice.

oneswig commented on drain.yml:
```diff
@@ -0,0 +1,20 @@
+  register: drained_nodes
+  until: "drained_nodes.stdout_lines is superset(nodes_to_drain)"
+  delay: 10
+  retries: 8640
```
Of course it is 8640 :-)

mkjpryor-stfc replied:
It would be really great if the timeout was user-configurable.

priteau replied:
It will be included in an updated commit.

oneswig commented on openhpc-repos.yml:
```diff
@@ -0,0 +1,37 @@
+      gpgcheck=1
+      gpgkey=http://build.openhpc.community/OpenHPC:/1.3:/Update{{ item }}/CentOS_7/repodata/repomd.xml.key
+      enabled=1
+  with_sequence: start=1 end=5 format=%d
```
tidy work!

oneswig commented on resume.yml:
```diff
@@ -0,0 +1,20 @@
+# Variables:
+# - nodes_to_resume: list of compute nodes to resume
+
+- name: Resume compute nodes
```
For idempotency, is there a means of checking the state before we change it to `RESUME`

priteau replied:
I am not sure I understand how it relates to idempotence. Did you mean that, for nodes that are already in DRAIN state (which could have been done by an admin), we should not resume them?

oneswig commented on update.yml:
```diff
@@ -0,0 +1,45 @@
+- hosts: login
+  gather_facts: no
+  tasks:
+  - import_tasks: resume.yml
```
As I believe @mkjpryor-stfc suggested elsewhere, we could have a common task file (eg, `slurm-scontrol.yml`) and pass in parameters for `drain` or `resume`.  If that fits and makes sense (it might not).

priteau replied:
I am not sure if there's much gain in this refactor. Anyway, this can be done in a separate pull request.

oneswig reviewed (commented):
Nice work @priteau just a couple of thoughts here and there, nothing major.

priteau committed:
Add support for updating packages on Slurm nodes

Slurm compute nodes are drained before applying updates. A reboot is
triggered if a new kernel has been installed.

This commit also includes code to configure older OpenHPC repos,
allowing to test an update of Slurm packages from 17.11.7 to 17.11.10
by passing configure_older_updates=true to the jasmin.yml playbook.

Note: the openhpc-repos.yml playbook is only for testing Slurm package
updates and can eventually be removed.

priteau commented:
Todo list for sprint 4:

- [x]  Use `needs-restarting` to check if reboot if needed
- [x]  Merge update.yml into the deploy template
- [x]  Look into draining and updating compute nodes one by one
- [x]  ~~Add option to install only security updates (see https://access.redhat.com/solutions/10021)~~

priteau commented:
CentOS doesn't provide the necessary metadata for `yum -y update --security` to work. It is only available for RHEL.

brtkwr commented:
I've changed the base to `awx-dev`

priteau committed:
Use needs-restarting instead of custom command

priteau committed:
Allow to update nodes independently

The number of compute nodes updated concurrently can be configured with
concurrent_compute_update.

priteau committed:
Rename loop variable to avoid conflict if nested

priteau commented:
- [x]  Extract update code so it can be applied to non-Slurm clusters
- [x] Restart services if required (check k8s repo)
- [x]  Test updates on NFS-only cluster
- [ ]  See what happens to Slurm when the NFS server is upgraded
- [x] Support variable `cluster_upgrade_system_packages` as per [Playbook compatibility](https://github.com/cedadev/jasmin-cluster-as-a-service/wiki/Playbook-compatibility#expected-variables)

priteau committed:
Improve package update code

* Check if updates are pending before doing anything else
* Restart services if required

priteau committed:
Integrate update code with main playbook

priteau committed:
Separate update code from Slurm drain/release

brtknr commented:
```
ok: [localhost] => {
    "openstack_stack": {
        "action": "UPDATE",
        "created_at": "2019-03-22T11:40:21",
        "creation_time": "2019-03-22T11:40:21",
        "deleted_at": null,
        "deletion_time": null,
        "description": "Top level Cluster-as-a-Service resource for the JASMIN Cloud.\n",
        "disable_rollback": true,
        "id": "bc1883b2-e1ea-4eb2-8f97-39a42c71642f",
        "identifier": "k8s/bc1883b2-e1ea-4eb2-8f97-39a42c71642f",
        "is_rollback_enabled": true,
        "location": {
            "cloud": "envvars",
            "project": {
                "domain_id": null,
                "domain_name": "jasmin",
                "id": "43d4c99f76d443a5a17b192e08fb2e30",
                "name": "caastest-U"
            },
            "region_name": "nova",
            "zone": null
        },
        "name": "k8s",
        "notification_topics": [],
        "outputs": [
            {
                "description": "No description given",
                "output_key": "info",
                "output_value": [
                    "",
                    ""
                ]
            },
            {
                "description": "No description given",
                "output_key": "node_groups",
                "output_value": [
                    {
                        "info": "",
                        "name": "kube-masters",
                        "nodes": [
                            {
                                "id": "bf07e06e-e6e4-486b-9fd0-aad716707507",
                                "ip": "192.171.139.249",
                                "name": "k8s-kube-masters-0"
                            }
                        ]
                    },
                    {
                        "info": "",
                        "name": "kube-workers",
                        "nodes": [
                            {
                                "id": "b6936242-6417-454f-91f1-ddb734fc8b77",
                                "ip": "192.168.3.176",
                                "name": "k8s-kube-workers-0"
                            },
                            {
                                "id": "7bda958f-b789-46cf-91a8-7a2fc008529b",
                                "ip": "192.168.3.175",
                                "name": "k8s-kube-workers-1"
                            },
                            {
                                "id": "da7e52af-b182-4a61-99f2-248530849554",
                                "ip": "192.168.3.178",
                                "name": "k8s-kube-workers-2"
                            }
                        ]
                    }
                ]
            },
            {
                "description": "No description given",
                "output_key": "deploy_private_key",
                "output_value": "..."
            }
        ],
        "owner": null,
        "parameters": {
            "OS::project_id": "43d4c99f76d443a5a17b192e08fb2e30",
            "OS::stack_id": "bc1883b2-e1ea-4eb2-8f97-39a42c71642f",
            "OS::stack_name": "k8s",
            "cluster_group_count": "2",
            "cluster_groups": "[{u'nodenet_fips': [{u'ip': u'192.171.139.249', u'uuid': u'8723c186-d15f-44d5-9421-e12bf2a03325'}], u'name': u'kube-masters', u'root_volume_size': 0, u'num_nodes': 1, u'nodenet_resource': u'Cluster::NodeNet1WithPreallocatedFIP', u'user': u'centos', u'flavor': u'j3.medium', u'node_resource': u'Cluster::Node'}, {u'name': u'kube-workers', u'root_volume_size': 0, u'num_nodes': 3, u'nodenet_resource': u'Cluster::NodeNet1', u'user': u'centos', u'flavor': u'j3.medium', u'node_resource': u'Cluster::Node'}]",
            "cluster_image": "centos-7-20190104",
            "cluster_keypair": "brtknr-33236e18e6285585a7f58f1006f88349",
            "cluster_network": "caastest-U-internal"
        },
        "parent": null,
        "properties": {
            "capabilities": []
        },
        "stack_name": "k8s",
        "stack_owner": null,
        "stack_status": "UPDATE_COMPLETE",
        "stack_status_reason": "Stack UPDATE completed successfully",
        "stack_user_project_id": "33deb6ddc73a41c7912722695a941c40",
        "status": "COMPLETE",
        "status_reason": "Stack UPDATE completed successfully",
        "tags": null,
        "tempate_description": "Top level Cluster-as-a-Service resource for the JASMIN Cloud.\n",
        "template_description": "Top level Cluster-as-a-Service resource for the JASMIN Cloud.\n",
        "timeout_mins": 60,
        "updated_at": "2019-03-22T12:36:43",
        "updated_time": "2019-03-22T12:36:43"
    }
}
```

oneswig commented on jasmin.yml:
```diff
@@ -64,6 +64,8 @@
         name: "https://github.com/openhpc/ohpc/releases/download/v1.3.GA/ohpc-release-1.3-1.el7.x86_64.rpm"
         state: present
 
+    - import_tasks: tasks/openhpc-repos.yml
```
As I was reading these changes I wondered if in future we should organise our files under an `openhpc/` subdirectory, and similarly for other deployments.  Just a thought for discussion at some point.

oneswig commented on tasks/openhpc-repos.yml:
```diff
@@ -0,0 +1,39 @@
+    value: 0
+  when: configure_older_updates is defined and configure_older_updates|bool
+
+- name: Enable repositories providing older updates
```
Too bad it appears we can't get hold of older versions of the openhpc-release rpm.

priteau replied:
We can get older ones, but their Slurm packages are completely different and not supported by our openhpc role.

oneswig reviewed (approved):
Looks good to me.

priteau committed:
Fix GlusterFS after reboot

priteau committed:
Apply security updates on Gluster clusters
